src/main/java/jenkins/scm/api/trait/SCMNavigatorRequest.java

```diff
@@ -0,0 +1,351 @@
+ * THE SOFTWARE.
+ */
+
+package jenkins.scm.api.trait;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.Closeable;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.SCMNavigator;
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.SCMSourceObserver;
+import jenkins.scm.impl.NoOpProjectObserver;
+
+/**
+ * Represents the context of an individual request for a call to
+ * {@link SCMNavigator#visitSources(SCMSourceObserver)} or an equivalent method.
+ *
+ * @since 2.2.0
+ */
+public abstract class SCMNavigatorRequest implements Closeable {
+
+    /**
+     * The {@link SCMNavigator} to use when applying the {@link #prefilters}.
+     */
+    @NonNull
+    private final SCMNavigator navigator;
+
+    /**
+     * The filters requiring context of the {@link SCMNavigatorRequest}, typically because the decision to filter may
+     * require making remote requests.
+     */
+    @NonNull
+    private final List<SCMSourceFilter> filters;
+
+    /**
+     * The filters that do not require context of the {@link SCMNavigatorRequest} and only require the {@link SCMNavigator}
+     * and {@code projectName} to decide exclusion - typically filtering based on the name.
+     */
+    @NonNull
+    private final List<SCMSourcePrefilter> prefilters;
+
+    /**
+     * The {@link SCMSourceObserver} for this request.
+     */
+    @NonNull
+    private final SCMSourceObserver observer;
+
+    /**
+     * The {@link SCMSourceObserver#getIncludes()} of {@link #observer}.
+     */
+    @CheckForNull
+    private final Set<String> observerIncludes;
+
+    /**
+     * The {@link SCMSourceTrait} instances to apply to {@link SCMSourceBuilder} instances.
+     */
+    @NonNull
+    private final List<SCMSourceTrait> traits;
+
+    /**
+     * The {@link SCMSourceDecorator} instances to apply to {@link SCMSourceBuilder} instances.
+     */
+    @NonNull
+    private final List<SCMSourceDecorator<?,?>> decorators;
+
+    /**
+     * Any {@link Closeable} objects that should be closed with the request.
+     */
+    // TODO widen type to AutoClosable once Java 7+
+    @NonNull
+    private final List<Closeable> managedClosables = new ArrayList<Closeable>();
+
+    /**
+     * Constructor.
+     *
+     * @param source  the source.
+     * @param context  the context.
+     * @param observer the observer.
+     */
+    protected SCMNavigatorRequest(@NonNull SCMNavigator source, @NonNull SCMNavigatorContext<?, ?> context,
+                                  @NonNull SCMSourceObserver observer) {
+        this.navigator = source;
+        this.filters = new ArrayList<SCMSourceFilter>(context.filters());
```

Comment:
because a context can be reused to service multiple requests and the requests may be being processes in parallel (not currently for the 1 implementation, but rather bake it into the API up front)